File: packages/python/src/a2a_lite/testing.py
"""
Testing utilities for A2A Lite agents.

Makes testing agents as simple as:

    from a2a_lite.testing import AgentTestClient

    def test_my_agent():
        client = AgentTestClient(agent)
        result = client.call("greet", name="World")
        assert result == "Hello, World!"
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Dict
from uuid import uuid4


@dataclass
class TestResult:
    """
    Structured result from a test client call.

    Provides multiple ways to access the result:
    - .data — parsed Python object (dict, list, int, str, etc.)
    - .text — raw text string
    - .json() — parse text as JSON (raises on invalid JSON)
    - .raw_response — the full A2A response dict
    """

    _data: Any
    _text: str
    raw_response: Dict[str, Any]

    @property
    def data(self) -> Any:
        """The parsed result value."""
        return self._data

    @property
    def text(self) -> str:
        """The raw text representation."""
        return self._text

    def json(self) -> Any:
        """Parse the text as JSON."""
        return json.loads(self._text)

    def __eq__(self, other: Any) -> bool:
        """Allow direct comparison with the data value for convenience."""
        if isinstance(other, TestResult):
            return self._data == other._data
        return self._data == other

    def __repr__(self) -> str:
        return f"TestResult(data={self._data!r})"


class AgentTestClient:
    """
    Simple test client for A2A Lite agents.

    Example:
        agent = Agent(name="Test", description="Test")

        @agent.skill("add")
        async def add(a: int, b: int) -> int:
            return a + b

        # In your test
        client = AgentTestClient(agent)
        assert client.call("add", a=2, b=3) == 5
    """

    def __init__(self, agent):
        """
        Create a test client for an agent.

        Args:
            agent: The A2A Lite Agent instance to test
        """
        self.agent = agent
        self._app = None
        self._client = None

    def _get_client(self):
        """Lazily create the test client."""
        if self._client is None:
            from starlette.testclient import TestClient as StarletteTestClient

            self._app = self.agent.get_app()
            self._client = StarletteTestClient(self._app)
        return self._client

    def call(self, skill: str, **params) -> Any:
        """
        Call a skill and return the result.

        Args:
            skill: Name of the skill to call
            **params: Parameters to pass to the skill

        Returns:
            The skill's return value (parsed from JSON if possible)

        Example:
            result = client.call("greet", name="World")
        """
        client = self._get_client()

        message = json.dumps({"skill": skill, "params": params})
        request_body = {
            "jsonrpc": "2.0",
            "method": "message/send",
            "id": uuid4().hex,
            "params": {
                "message": {
                    "role": "user",
                    "parts": [{"type": "text", "text": message}],
                    "messageId": uuid4().hex,
                }
            },
        }

        response = client.post("/", json=request_body)
        response.raise_for_status()
        data = response.json()

        # Extract the actual result from A2A response
        return self._extract_result(data)

    def _extract_result(self, response: Dict) -> TestResult:
        """Extract the skill result from A2A response."""
        if "error" in response:
            raise TestClientError(response["error"])

        result = response.get("result", {})

        # Get text from message parts
        parts = result.get("parts", [])
        for part in parts:
            if part.get("kind") == "text" or part.get("type") == "text":
                text = part.get("text", "")
                # Try to parse as JSON
                try:
                    data = json.loads(text)
                except json.JSONDecodeError:
                    data = text
                return TestResult(_data=data, _text=text, raw_response=response)

        return TestResult(_data=result, _text=json.dumps(result), raw_response=response)

    def get_agent_card(self) -> Dict[str, Any]:
        """
        Fetch the agent card.

        Returns:
            The agent card as a dictionary
        """
        client = self._get_client()
        response = client.get("/.well-known/agent.json")
        response.raise_for_status()
        return response.json()

    def list_skills(self) -> list[str]:
        """
        Get list of available skill names.

        Returns:
            List of skill names
        """
        card = self.get_agent_card()
        return [s.get("name", s.get("id")) for s in card.get("skills", [])]

    def stream(self, skill: str, **params) -> list[Any]:
        """
        Call a streaming skill and collect all results.

        Args:
            skill: Name of the skill to call
            **params: Parameters to pass to the skill

        Returns:
            List of all streamed values

        Example:
            results = client.stream("count", limit=3)
            assert len(results) == 3
        """
        import asyncio

        # Access skills directly from agent
        skill_def = self.agent._skills.get(skill)

        if not skill_def:
            raise TestClientError(f"Unknown skill: {skill}")

        # Call handler directly and collect results
        results = []

        async def run_handler():
            handler = skill_def.handler
            gen = handler(**params)

            # Handle both async and sync generators
            if hasattr(gen, "__anext__"):
                async for value in gen:
                    results.append(value)
            elif hasattr(gen, "__next__"):
                for value in gen:
                    results.append(value)
            else:
                # Not a generator, just a coroutine
                result = await gen
                results.append(result)

        # Handle both sync and async calling contexts
        try:
            asyncio.get_running_loop()
            # Already in an async context — run in a separate thread
            import concurrent.futures

            with concurrent.futures.ThreadPoolExecutor(1) as pool:
                pool.submit(asyncio.run, run_handler()).result()
        except RuntimeError:
            # No running loop — safe to use asyncio.run()
            asyncio.run(run_handler())
        return results


class TestClientError(Exception):
    """Error from test client."""

    pass


# Async version for async tests
class AsyncAgentTestClient:
    """
    Async test client for A2A Lite agents.

    Example:
        async def test_my_agent():
            client = AsyncAgentTestClient(agent)
            result = await client.call("greet", name="World")
            assert result == "Hello, World!"
    """

    def __init__(self, agent):
        self.agent = agent
        self._app = None
        self._client = None

    async def _get_client(self):
        """Lazily create the async test client."""
        if self._client is None:
            import httpx

            self._app = self.agent.get_app()
            self._client = httpx.AsyncClient(
                app=self._app, base_url="http://testserver"
            )
        return self._client

    async def call(self, skill: str, **params) -> Any:
        """
        Call a skill and return the result.

        Args:
            skill: Name of the skill to call
            **params: Parameters to pass to the skill

        Returns:
            The skill's return value
        """
        client = await self._get_client()

        message = json.dumps({"skill": skill, "params": params})
        request_body = {
            "jsonrpc": "2.0",
            "method": "message/send",
            "id": uuid4().hex,
            "params": {
                "message": {
                    "role": "user",
                    "parts": [{"type": "text", "text": message}],
                    "messageId": uuid4().hex,
                }
            },
        }

        response = await client.post("/", json=request_body)
        response.raise_for_status()
        data = response.json()

        return self._extract_result(data)

    def _extract_result(self, response: Dict) -> TestResult:
        """Extract the skill result from A2A response."""
        if "error" in response:
            raise TestClientError(response["error"])

        result = response.get("result", {})
        parts = result.get("parts", [])

        for part in parts:
            if part.get("kind") == "text" or part.get("type") == "text":
                text = part.get("text", "")
                try:
                    data = json.loads(text)
                except json.JSONDecodeError:
                    data = text
                return TestResult(_data=data, _text=text, raw_response=response)

        return TestResult(_data=result, _text=json.dumps(result), raw_response=response)

    async def close(self):
        """Close the client."""
        if self._client:
            await self._client.aclose()
            self._client = None

File: packages/typescript/src/testing.ts
/**
 * Testing utilities for A2A Lite TypeScript.
 *
 * Makes testing agents as simple as:
 *
 *   import { AgentTestClient } from 'a2a-lite';
 *
 *   const client = new AgentTestClient(agent);
 *   const result = await client.call("greet", { name: "World" });
 *   expect(result).toBe("Hello, World!");
 */

import type { Agent } from './agent.js';
import type { AgentCard } from '@a2a-js/sdk';
import type { Request, Response, NextFunction } from 'express';
import type { SkillDefinition } from './types.js';

/**
 * Internal interface used to access private Agent fields in tests
 * without resorting to `as any`.
 */
interface AgentInternals {
  skills: Map<string, SkillDefinition>;
}

/** Type guard: checks if a value is async iterable. */
function isAsyncIterable(value: unknown): value is AsyncIterable<unknown> {
  return value !== null && typeof value === 'object' && Symbol.asyncIterator in (value as object);
}

/** Type guard: checks if a value is (sync) iterable. */
function isIterable(value: unknown): value is Iterable<unknown> {
  return value !== null && typeof value === 'object' && Symbol.iterator in (value as object);
}

/**
 * Structured result from a test client call.
 *
 * Provides multiple ways to access the result:
 * - .data — parsed JavaScript value (object, array, number, string, etc.)
 * - .text — raw text string
 * - .json() — parse text as JSON (throws on invalid JSON)
 */
export class TestResult {
  private _data: unknown;
  private _text: string;

  constructor(data: unknown, text: string) {
    this._data = data;
    this._text = text;
  }

  get data(): unknown {
    return this._data;
  }

  get text(): string {
    return this._text;
  }

  json(): unknown {
    return JSON.parse(this._text);
  }

  toString(): string {
    return `TestResult(${JSON.stringify(this._data)})`;
  }
}

export class AgentTestClient {
  private agent: Agent;
  private baseUrl: string;

  constructor(agent: Agent, options?: { baseUrl?: string }) {
    this.agent = agent;
    this.baseUrl = options?.baseUrl ?? 'http://localhost:8787';
  }

  /**
   * Call a skill and return the result.
   */
  async call(skill: string, params: Record<string, unknown> = {}): Promise<TestResult> {
    const app = this.agent.buildApp();

    // Create A2A v1.0 JSON-RPC request
    const message = JSON.stringify({ skill, params });
    const requestBody = {
      jsonrpc: '2.0',
      method: 'SendMessage',
      id: Math.random().toString(36).slice(2),
      params: {
        message: {
          role: 'ROLE_USER',
          messageId: Math.random().toString(36).slice(2),
          parts: [{ text: message }],
        },
      },
    };

    // Use a promise-based approach to invoke the route handler
    return new Promise((resolve, reject) => {
      const mockReq = {
        body: requestBody,
        method: 'POST',
        url: '/',
        headers: {
          'content-type': 'application/json',
          'a2a-version': '1.0',
        },
        hostname: 'localhost',
        get: (header: string) => {
          if (header.toLowerCase() === 'host') return 'localhost:8787';
          if (header.toLowerCase() === 'content-type') return 'application/json';
          if (header.toLowerCase() === 'a2a-version') return '1.0';
          return undefined;
        },
      };

      const mockRes = {
        statusCode: 200,
        headers: {} as Record<string, string>,
        setHeader: (name: string, value: string) => {
          mockRes.headers[name.toLowerCase()] = value;
        },
        getHeader: (name: string) => mockRes.headers[name.toLowerCase()],
        status: (code: number) => {
          mockRes.statusCode = code;
          return mockRes;
        },
        json: (data: Record<string, unknown>) => {
          try {
            const result = this.extractResult(data);
            resolve(result);
          } catch (err) {
            reject(err);
          }
        },
        send: (data: string) => {
          try {
            const parsed = JSON.parse(data);
            const result = this.extractResult(parsed);
            resolve(result);
          } catch {
            console.debug('Response is not JSON, returning raw text');
            resolve(new TestResult(data, data));
          }
        },
        end: () => {
          resolve(new TestResult(undefined, ''));
        },
      };

      // Handle the request through Express
      const next: NextFunction = (err?: unknown) => {
        if (err) reject(err instanceof Error ? err : new Error(String(err)));
      };
      app(mockReq as unknown as Request, mockRes as unknown as Response, next);
    });
  }

  /**
   * Call a streaming skill and collect all results.
   */
  async stream(skill: string, params: Record<string, unknown> = {}): Promise<unknown[]> {
    // Access skill directly for streaming tests via typed internal interface
    const skills = (this.agent as unknown as AgentInternals).skills;
    const skillDef = skills.get(skill);

    if (!skillDef) {
      throw new TestClientError(`Unknown skill: ${skill}`);
    }

    const results: unknown[] = [];
    const gen = skillDef.handler(params);

    if (isAsyncIterable(gen)) {
      for await (const value of gen) {
        results.push(value);
      }
    } else if (isIterable(gen)) {
      for (const value of gen) {
        results.push(value);
      }
    } else {
      results.push(await gen);
    }

    return results;
  }

  /**
   * Get the agent card.
   */
  getAgentCard(): AgentCard {
    return this.agent.buildAgentCard();
  }

  /**
   * List available skills.
   */
  listSkills(): string[] {
    const card = this.getAgentCard();
    return card.skills?.map((s) => s.name) ?? [];
  }

  /**
   * Extract result from A2A v1.0 JSON-RPC response.
   * The result envelope is `{ message: {...} }` or `{ task: {...} }`;
   * text parts use the `{ "text": ... }` wire shape.
   */
  private extractResult(response: Record<string, unknown>): TestResult {
    if (response.error) {
      throw new TestClientError(JSON.stringify(response.error));
    }

    const envelope = (response.result ?? {}) as Record<string, unknown>;
    const result = (envelope.message ?? envelope.task ?? envelope) as Record<string, unknown>;

    // Handle A2A message format
    if (result?.parts) {
      const parts = result.parts as Array<{ text?: string }>;
      const textPart = parts.find((p) => typeof p.text === 'string');
      if (textPart?.text) {
        let data: unknown;
        try {
          data = JSON.parse(textPart.text);
        } catch {
          console.debug('Result text is not JSON, returning as string');
          data = textPart.text;
        }
        return new TestResult(data, textPart.text);
      }
    }

    const text = JSON.stringify(result);
    return new TestResult(result, text);
  }
}

export class TestClientError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TestClientError';
  }
}

/**
 * Async-explicit test client — same as AgentTestClient.
 *
 * Provided for API parity with the Python SDK's AsyncAgentTestClient.
 * In TypeScript everything is already async, so this is an alias.
 */
export class AsyncAgentTestClient extends AgentTestClient {}
